lib/remap-lang/src/expression/function.rs

```diff
@@ -0,0 +1,96 @@
+use super::Error as E;
+use crate::function::Split;
+use crate::{Argument, ArgumentList, Expression, Function as _, Object, Result, State, Value};
+
+#[derive(thiserror::Error, Debug, PartialEq)]
+pub enum Error {
+    #[error("undefined")]
+    Undefined,
+
+    #[error("invalid argument count (expected at most {0}, got {0})")]
+    Arity(usize, usize),
+
+    #[error(r#"unknown argument keyword "{0}""#)]
+    Keyword(String),
+
+    #[error(r#"missing required argument "{0}" (position {1})"#)]
+    Required(String, usize),
+}
+
+#[derive(Debug)]
+pub(crate) struct Function {
+    function: Box<dyn Expression>,
+}
+
+impl Function {
+    pub(crate) fn new(ident: String, arguments: Vec<(Option<String>, Argument)>) -> Result<Self> {
+        // TODO: move to `Runtime` so that we can eventually expose
+        // `register_function`.
+        let functions = vec![Split];
+
+        let function = functions
+            .iter()
+            .find(|b| b.identifier() == ident)
+            .ok_or_else(|| E::Function(ident.clone(), Error::Undefined))?;
```

Comment:
Is it safe to presume this will become a map at some point? It's only evaluated at load time, so not a big deal.

Reply:
Yeah, I agree. This will change in a follow-up PR when the Vec is pulled out from here into the `Program` struct.